util/file.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package util
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/keybase/go-logging"
+)
+
+// File uses a safer file API
+type File struct {
+	name string
+	data []byte
+	perm os.FileMode
+}
+
+// SafeWriter defines a writer that is safer (atomic)
+type SafeWriter interface {
+	GetFilename() string
+	WriteTo(io.Writer) (int64, error)
+}
+
+// NewFile returns a File
+func NewFile(name string, data []byte, perm os.FileMode) File {
+	return File{name, data, perm}
+}
+
+// Save file
+func (f File) Save(log logging.Logger) error {
+	return SafeWriteToFile(f, f.perm, log)
+}
+
+// GetFilename returns the file name for SafeWriter
+func (f File) GetFilename() string {
+	return f.name
+}
+
+// WriteTo is for SafeWriter
+func (f File) WriteTo(w io.Writer) (int64, error) {
+	n, err := w.Write(f.data)
+	return int64(n), err
+}
+
+// SafeWriteToFile to safely write to a file. Use mode=0 for default permissions.
+func SafeWriteToFile(t SafeWriter, mode os.FileMode, log logging.Logger) error {
+	fn := t.GetFilename()
+	log.Debugf("Writing to %s", fn)
+	tmpfn, tmp, err := OpenTempFile(fn, "", mode)
+	log.Debugf("Temporary file generated: %s", tmpfn)
+	if err != nil {
+		return err
+	}
+
+	_, err = t.WriteTo(tmp)
+	if err == nil {
```

Comment:
this would be better laid out as follows:
```go
if err != nil {
    log.Errorf("Error writing temporary file %s: %s", tmpfn, err)
    _ = tmp.Close()
    _ = os.Remove(tmpfn)
    return err
}
err = tmp.Close()
if err != nil {
    log.Errorf("Error closing temporary file %s: %s", tmpfn, err)
    _ = os.Remove(tmpfn)
    return err
}
log.Debugf("Wrote to %s", fn)
return os.Rename(tmpfn, fn) // TODO: rollback if this fails
```